pennylane/interfaces/execution.py

```diff
@@ -339,6 +363,7 @@ def cost_fn(params, x):
                 cache,
                 pass_kwargs=True,
                 return_tuple=False,
+                finite_shots=device.shots is not None,

```

Comment:
@dime10 just a quick flyby comment: the execution `fn` that is being passed to `cache_execute` is being wrapped in a `set_shots` decorator, that potentially overrides the device shots, so this should be taken into account.

Rather than using a new `finite_shots` argument, you may be able to extract the shot information from `fn` _inside_ the `cache_execute` function. This might be quite a clean implementation.

Reply:
Thanks for the tip!

Reply:
@josh146 A bit hacky, but the new version now inspects the set_shots context manager to peak at the shots device attribute.